velox/functions/sparksql/String.h

```diff
@@ -845,4 +845,103 @@ struct TranslateFunction {
   }
 };
 
+/// Decodes `string` in 'application/x-www-form-urlencoded' format using UTF-8
+/// encoding scheme.
+template <typename T>
+struct UrlDecoderFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // ASCII input does NOT always produce ASCII result.
+  static constexpr bool is_default_ascii_behavior = false;
+
+  int decode(StringView input) {
+    assert(input.size() == 2);
+    auto toIntFromHex = [](char x) {
+      switch (x) {
+        case '0':
+          return 0;
+        case '1':
+          return 1;
+        case '2':
+          return 2;
+        case '3':
+          return 3;
+        case '4':
+          return 4;
+        case '5':
+          return 5;
+        case '6':
+          return 6;
+        case '7':
+          return 7;
+        case '8':
+          return 8;
+        case '9':
+          return 9;
+        case 'a':
+        case 'A':
+          return 10;
+        case 'b':
+        case 'B':
+          return 11;
+        case 'c':
+        case 'C':
+          return 12;
+        case 'd':
+        case 'D':
+          return 13;
+        case 'e':
+        case 'E':
+          return 14;
+        case 'f':
+        case 'F':
+          return 15;
+        default:
+          VELOX_USER_FAIL(
+              "Unexpected char for converting hex digit: {}!",
+              std::string(1, x));
+      }
+    };
+
+    return (toIntFromHex(input.data()[0]) << 4) |
+        (toIntFromHex(input.data()[1]));
+  }
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& input) {
+    if (input.size() == 0) {
+      result.setEmpty();
+      return;
+    }
+    auto size = input.size();
+    result.reserve(size);
+    int index = 0;
+    int i = 0;
+    while (i < size) {
+      if (input.data()[i] == '+') {
+        result.data()[index++] = ' ';
+        i++;
+      } else if (input.data()[i] == '%') {
+        while ((i + 2) < size && input.data()[i] == '%') {
```

Comment:
https://github.com/facebookincubator/velox/blob/main/velox/functions/lib/string/StringImpl.h#L334
You can get the char from input easily like that.